pkg/odp/segment_api_manager.go

```diff
@@ -0,0 +1,229 @@
+	requester utils.Requester
+}
+
+// NewSegmentAPIManager creates and returns a new instance of DefaultSegmentAPIManager.
+func NewSegmentAPIManager(requester utils.Requester) *DefaultSegmentAPIManager {
+	if requester == nil {
+		requester = utils.NewHTTPRequester(logging.GetLogger("", "SegmentAPIManager"))
+	}
+	return &DefaultSegmentAPIManager{requester: requester}
+}
+
+// FetchSegments returns qualified ODP segments
+func (s *DefaultSegmentAPIManager) FetchSegments(apiKey, apiHost, userKey, userValue string, segmentsToCheck []string) ([]string, error) {
+
+	// Creating query for odp request
+	requestQuery := s.createRequestQuery(userKey, userValue, segmentsToCheck)
+
+	// Creating request
+	apiEndpoint := apiHost + "/v3/graphql"
+	headers := []utils.Header{{Name: "Content-Type", Value: "application/json"}, {Name: "x-api-key", Value: apiKey}}
+
+	// handling edge cases
+	response, _, _, err := s.requester.Post(apiEndpoint, requestQuery, headers...)
+	if err != nil {
+		return nil, fmt.Errorf(fetchSegmentsFailedError, err.Error())
+	}
+
+	// Checking if response is decodable
+	responseMap := map[string]interface{}{}
+	if err = json.Unmarshal(response, &responseMap); err != nil {
+		return nil, fmt.Errorf(fetchSegmentsFailedError, "decode error")
+	}
+
+	// most meaningful ODP errors are returned in 200 success JSON under {"errors": ...}
+	if odpErrors, ok := s.extractComponent("errors", responseMap).([]interface{}); ok {
+		if odpError, ok := odpErrors[0].(map[string]interface{}); ok {
+			if errorClass, ok := s.extractComponent("extensions.classification", odpError).(string); ok {
+				if errorClass == "InvalidIdentifierException" {
+					return nil, errors.New(invalidSegmentIdentifier)
+				}
+				return nil, fmt.Errorf(fetchSegmentsFailedError, errorClass)
+			}
+		}
+		return nil, fmt.Errorf(fetchSegmentsFailedError, "decode error")
+	}
+
+	// Retrieving audience edges from response
+	audienceDictionaries, ok := s.extractComponent("data.customer.audiences.edges", responseMap).([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(fetchSegmentsFailedError, "decode error")
+	}
+
+	// Parsing and returning qualified segments
+	returnSegments := []string{}
+	for _, audDict := range audienceDictionaries {
+		convertedAudDict, ok := audDict.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		if jsonbody, err := json.Marshal(convertedAudDict["node"]); err == nil {
+			var audience Audience
+			if err := json.Unmarshal(jsonbody, &audience); err != nil {
+				continue
+			}
+			if audience.isQualified() {
+				returnSegments = append(returnSegments, audience.Name)
+			}
+		}
+	}
+	return returnSegments, nil
+}
+
+// Creates graphql query
+func (s DefaultSegmentAPIManager) createRequestQuery(userKey, userValue string, segmentsToCheck []string) map[string]interface{} {
+	query := fmt.Sprintf(`query($userId: String, $audiences: [String]) {customer(%s: $userId) {audiences(subset: $audiences) {edges {node {name state}}}}}`, userKey)
```

Comment:
you don't need fmt.Sprintf. 
Please use multiline with indentation to have more visibility.

Reply:
String interpolation is only supported using fmt.Sprintf, will still look into it.